File: frida/gttrace/server/lib/mod_lookup.py
"""Module range indexing for address-to-module resolution."""

from bisect import bisect_left, bisect_right
from dataclasses import dataclass

@dataclass(frozen=True)
class ModRVA:
    """Identify an address by module name and relative virtual address (RVA)."""
    mod: str
    rva: int

class ModLookup:
    """Interval map for mapping absolute addresses to module RVAs.

    The lookup table keeps parallel sorted lists for module start/end
    addresses, enabling O(log n) lookups via bisect.
    """
    def __init__(self):
        self.starts = []
        self.ends = []
        self.names = []

    def add(self, start: int, end: int, name: str):
        """Register a new module range by address.

        Raises:
            ValueError: If the new range overlaps existing ranges.
        """
        assert start < end
        i = bisect_left(self.starts, start)

        if i > 0 and start < self.ends[i-1]:
            raise ValueError("overlaps previous interval")

        if i < len(self.starts) and end > self.starts[i]:
            raise ValueError("overlaps next interval")

        self.starts.insert(i, start)
        self.ends.insert(i, end)
        self.names.insert(i, name)

    def rem(self, start: int, end: int):
        """Remove a module range if it matches exactly."""
        i = bisect_left(self.starts, start)
        if i >= len(self.starts) or self.starts[i] != start or self.ends[i] != end:
            return False
        self.starts.pop(i); self.ends.pop(i); self.names.pop(i)
        return True

    def lookup(self, addr: int):
        """Lookup an address and return a ModRVA if it falls within a module."""
        i = bisect_right(self.starts, addr) - 1
        if i >= 0 and addr < self.ends[i]:
            rva = addr - self.starts[i]
            return ModRVA(self.names[i], rva)
        return None

File: frida/gttrace/server/src/mod_lookup.rs
//! Module interval lookup helper for resolving addresses to module RVAs.

/// A resolved module name with its relative virtual address (RVA).
#[derive(Debug, Clone)]
pub struct ModRva {
    /// Module name the address belongs to.
    pub module: String,
    /// Address offset from module base.
    pub rva: u64,
}

/// Ordered module interval table for fast address-to-module lookups.
#[derive(Debug, Default, Clone)]
pub struct ModLookup {
    starts: Vec<u64>,
    ends: Vec<u64>,
    names: Vec<String>,
}

impl ModLookup {
    /// Create a new, empty lookup table.
    pub fn new() -> Self { Self::default() }

    /// Insert a non-overlapping module range.
    pub fn add(&mut self, start: u64, end: u64, name: String) -> anyhow::Result<()> {
        anyhow::ensure!(start < end);

        let i = self.starts.partition_point(|&x| x < start);

        if i > 0 && start < self.ends[i - 1] {
            anyhow::bail!("overlaps previous interval");
        }
        if i < self.starts.len() && end > self.starts[i] {
            anyhow::bail!("overlaps next interval");
        }

        self.starts.insert(i, start);
        self.ends.insert(i, end);
        self.names.insert(i, name);
        Ok(())
    }

    /// Remove a module range only if the start and end match exactly.
    pub fn rem_exact(&mut self, start: u64, end: u64) -> bool {
        let i = self.starts.partition_point(|&x| x < start);
        if i >= self.starts.len() || self.starts[i] != start || self.ends[i] != end {
            return false;
        }
        self.starts.remove(i);
        self.ends.remove(i);
        self.names.remove(i);
        true
    }

    /// Resolve an absolute address into a module name and RVA.
    pub fn lookup(&self, addr: u64) -> Option<ModRva> {
        let i = self.starts.partition_point(|&x| x <= addr);
        if i == 0 {
            return None;
        }
        let idx = i - 1;
        if addr < self.ends[idx] {
            Some(ModRva {
                module: self.names[idx].clone(),
                rva: addr - self.starts[idx],
            })
        } else {
            None
        }
    }
}
